text/2018-10-25-tikv-client-rust.md

```diff
@@ -0,0 +1,395 @@
+
+All structures and functions will otherwise follow the [Rust API Guidelines](https://rust-lang-nursery.github.io/api-guidelines/), some of which will be enforced by `clippy`.
+
+## Installation
+
+To utilize the client programmatically, users will be able to add the `tikv-client` crate to their `Cargo.toml`'s dependencies. Then they must use the crate with `use tikv_client;`. Unfortunately due to Rust’s naming conventions this inconsistency is in place.
+
+To utilize the command line client, users will be able to install the binary via `cargo install tikv-client`. They will then be able to access the client through the binary `tikv-client`. If they wish for a different name they can alias it in their shell.
+
+## Usage
+
+## Two types of APIs
+TiKV provides two types of APIs for developers:
+- The Raw Key-Value API
+
+    If your application scenario does not need distributed transactions or MVCC (Multi-Version Concurrency Control) and only needs to guarantee the atomicity towards one key, you can use the Raw Key-Value API.
+
+- The Transactional Key-Value API
+
+    If your application scenario requires distributed ACID transactions and the atomicity of multiple keys within a transaction, you can use the Transactional Key-Value API.
+
+Generally, the Raw Key-Value API has higher throughput and lower latency compared to the Transactional Key-Value API. If distributed ACID transactions are not required, Raw Key-Value API is preferred over Transactional Key-Value API for better performance and ease of use.
+
+The client provides two types of APIs in two separate modules for developers to choose from.
+
+### The common data types
+
+- Key: raw binary data
+- Value: raw binary data
+- KvPair: Key-value pair type
+- KeyRange: Half-open interval of keys
+- Config: Configuration for client
+
+### Raw Key-Value API Basic Usage
+
+To use the Raw Key-Value API, take the following steps:
+
+1. Create an instance of Config to specify endpoints of PD (Placement Driver) and optional security config
+
+    ```rust
+        let config = Config::new(vec!["127.0.0.1:2379"]);
+    ```
+
+2. Create a Raw Key-Value client.
+
+    ```rust
+        let client = RawClient::new(&config);
+    ```
+
+3. Call the Raw Key-Value client methods to access the data on TiKV. The Raw Key-Value API contains following methods
+
+    ```rust
+        fn get<K, C>(&self, key: K, cf: C) -> KvFuture<Value>
+        where
+            K: AsRef<Key>,
+            C: Into<Option<String>>;
+
+        fn batch_get<K, C>(&self, keys: K, cf: C) -> KvFuture<Vec<KvPair>>
+        where
+            K: AsRef<[Key]>,
+            C: Into<Option<String>>;
+
+        fn put<P, C>(&self, pair: P, cf: C) -> KvFuture<()>
+        where
+            P: Into<KvPair>,
+            C: Into<Option<String>>;
+
+        fn batch_put<I, P, C>(&self, pairs: I, cf: C) -> KvFuture<()>
```

Comment:
Good idea. It definitely looks better than many single letter type variables.

Reply:
BTW: The inlining style you mentioned? Did you mean this?
```rust
    fn batch_put<
        IntoIter: IntoIterator<Item = Pair>,
        Pair: Into<KvPair>,
        ColumnFamily: Into<Option<String>>,
    >(  
        &self,
        _pairs: Pair,
        _cf: ColumnFamily,
    ) -> KvFuture<()>;
```
Using trait bound directly as argument type is prohibited.

Reply:
Oh, I see what you mean now. It would be something like this
```rust
    fn batch_put(                                                                                                                                              
        &self,                                                                                                                                                 
        pairs: impl IntoIterator<Item = impl Into<KvPair>>,                                                                                                    
        cf: impl Into<Option<String>>,                                                                                                                         
    ) -> KvFuture<()>;
```
It's the first time I realize the reason why impl trait exists for arguments. It's definitely easier to read than where syntax. Thanks.

Reply:
Sadly, this particular case isn't supported.
```rust
    fn batch_scan(
        &self,
        ranges: impl AsRef<[impl RangeBounds<Key>]>,
        each_limit: u32,
        key_only: bool,
        cf: impl Into<Option<String>>,
    ) -> KvFuture<Vec<KvPair>>;
```
I'm going to leave this one with the old generic syntax and change all other methods to impl trait syntax.

Reply:
And this way
```rust
    fn batch_scan<Ranges: AsRef<[Bounds]>, Bounds: RangeBounds<Key>, CF: Into<Option<String>>>(                                                                
        &self,
        ranges: Ranges,
        each_limit: u32,
        key_only: bool,
        cf: CF,
    ) -> KvFuture<Vec<KvPair>>;  
```
looks less than where style for me
```rust
    fn batch_scan<Ranges, Bounds, CF>(
        &self,
        ranges: Ranges,
        each_limit: u32,
        key_only: bool,
        cf: CF,
    ) -> KvFuture<Vec<KvPair>>
    where
        Ranges: AsRef<[Bounds]>,
        Bounds: RangeBounds<Key>,
        CF: Into<Option<String>>;
```
What do you think? @gregwebs

Reply:
Great! I agree on `batch_scan`. I am not against `where` at all, just against the common practice of using one letter variable names, particularly when they are not 100% consistent in their meaning.

Reply:
It's somewhat like this now
```rust
pub struct BatchScan<'a, AClient: Client>;

impl<'a, AClient: Client> BatchScan<'a, AClient> {
    fn new(client: &'a AClient, ranges: Vec<(Key, Key)>, each_limit: u32) -> Self;
    pub fn key_only(mut self) -> Self; 
    pub fn cf(mut self, cf: impl Into<ColumnFamily>) -> Self;
}

impl<'a, AClient: Client> Future for BatchScan<'a, AClient> {
    type Item = Vec<KvPair>;
    type Error = (); 
    fn poll(&mut self) -> Poll<Self::Item, Self::Error>;
}

pub trait Client {
    type AClient: Client;

    fn batch_scan<Ranges, Bounds>(
        &self,
        ranges: Ranges,
        each_limit: u32,
    ) -> BatchScan<Self::AClient>
    where
        Ranges: AsRef<[Bounds]>,
        Bounds: RangeBounds<Key>;
}
``` 
You can find the complete interface at [here](https://github.com/tikv/client-rust/blob/public_api_rfc/src/raw.rs)

PLTA @gregwebs

Reply:
Hmmm, I don't think the `AClient: Client` reads very well. In this case since it's pretty clear we expect a `Client` (there aren't multiple bounds, and the trait name is very clear) so I think just `C: Client` is sufficient. This is fairly idiomatic for Rust code.

Reply:
Notably this looks strange as the associated type in `trait Client { type ... }`

Reply:
Makes sense, let's change it first and see how it looks.

Reply:
Oh, just want to double check if you meant only the AClient: Client inside Client trait. 
There are quite some AClient: Client out there.

Reply:
I think the `AFoo: Foo` semantic in general is not something I've seen much in Rust, and it reads quite strange.

Reply:
I changed them to Impl: Client. Maybe that looks better and Impl is widely used to represent an implementation of something which could be appropriate for this scenario.